Added support for missing clear command

Ifera opened the pull request:
## Introduction
<!-- Explain existing problems or why this pull request is necessary -->
This PR aims to provide support for vanilla command `clear`. With another [PR](https://github.com/pmmp/Language/pull/58) to `pmmp/Language` repo adding support for the missing strings. Exact vanilla-like functionality was achieved.

### Relevant issues
<!-- List relevant issues here -->
<!--

* Fixes #1
* Fixes #2

-->
* Fixes #3840 

## Changes
- Added `/clear` command
- Registered the command to command map
- Registered permissions for the command
### API changes
<!-- Any additions to the API that should be documented in release notes? -->
`N/A`
### Behavioural changes
<!-- Any change in how the server behaves, or its performance? -->
`N/A`
## Backwards compatibility
<!-- Any possible backwards incompatible changes? How are they solved, or how can they be solved? -->
`Yes`
## Follow-up
<!-- Suggest any actions to be done before/after merging this pull request -->
<!--

Requires translations:

| Name | Value in eng.ini |
| :--: | :---: |
| `foo.bar` | `Foo bar` |

-->
`N/A` 
_(translations already added as explained above)_
## Tests
<!--
Details should be provided of tests done. Simply saying "tested" or equivalent is not acceptable.

Attach scripts or actions to test this pull request, as well as the result
-->
Not quite sure what to add in tests, anyways here are some screenshots in [this gist](https://gist.github.com/Ifera/c6c5d39b99321a148c42fde62c80d737).

Commit:
implemented vanilla command: /clear

Commit:
register the command in the command map

Commit:
register permissions for clear command

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,178 @@
+		}
+
+		if($cleared > 0){
+			$sender->sendMessage(new TranslationContainer("%commands.clear.success", [$target->getName(), $cleared]));
```
This should use `Command::broadcastCommandMessage()` for auditing.

dktapps replied:
This should use `Command::broadcastCommandMessage()` for auditing.

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,178 @@
+			}
+
+			//vanilla behaviour, removes item from armor inventory, regardless of the count specified
+			if(($index = $target->getArmorInventory()->first($item)) !== -1){
```
I don't like this. This behaviour looks buggy and it doesn't make sense to replicate vanilla bugs.

Ifera replied:
So what do you suggest as a replacement..? Just check if the item is

dktapps replied:
I think a more sensible alternative would be to check all the available inventories for the item in a specific order. Maybe the armour inventory could be prioritized, but it doesn't make sense to remove items from multiple inventories when only a count of one was desired.

Ifera replied:
Alright, seems fair :+1:

Ifera replied:
Fixed, please review.

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,178 @@
+		}
+
+		//checking players inventory for all the items matching the criteria
+		if($item !== null and $item->getMeta() === 0 and $maxCount === 0){
```
why is there a `getMeta()` check here?

Ifera replied:
Vanilla `clear` command does more than just simply clearing the inventory of the player. Vanilla syntax is, 
```clear [player] [item name] [data] [max count]```
Where item being the item name and data being the damage/meta. If the meta and the max count are set to zero then it reports back all the items matching the items specified in the players inventory. 
More info [here](https://www.digminecraft.com/game_commands/clear_command.php) or see the attached screenshot.

![image](https://user-images.githubusercontent.com/32965703/95328052-68cce900-08be-11eb-9028-3a30bfe57acf.png)

dktapps replied:
This seems out of character for a `/clear` command...

Ifera replied:
True but that's how its done in vanilla... Just to be sure, I checked this in-game as well.

dktapps replied:
OK, so /clear has some behaviour that's not relevant to clearing when using count zero. Fine. But this still doesn't explain the metadata zero check.

dktapps replied:
EX: `/clear steve bucket 8 0` has the same behaviour as `/clear steve bucket 0 0`.

Ifera replied:
Ah true, forgot about that.

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,178 @@
+
+			$target = $sender;
+		}else{
+			if(count($args) < 1){
```
This can be reduced to `elseif`.

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,178 @@
+			}
+
+			return true;
+		}
```
I don't understand the purpose of this entire block. The point of `/clear` is to remove items.

Ifera replied:
Discussed in above comment.

Commit:
use `Command::broadcastCommandMessage()`

Commit:
remove `getMeta()` check

Commit:
reduce to `elseif`

Commit:
prioritize clearing of armor inventory (and fix vanilla logic)

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,182 @@
+				$maxCount = -1;
+			}
+
+			if($item !== null){
```
This will always be true (presence of args[2] guarantees the presence of args[1], which is the item type).

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,182 @@
+		if(isset($args[2])){
+			if(!is_numeric($args[2])){
+				$maxCount = -1;
+			}
```
This code is useless (because the condition below is always true).

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,182 @@
+		}
+
+		$target = null;
+		if($sender instanceof Player){
```
It doesn't make sense to check this before checking if args[0] is set. A more intuitive logic flow would be
```
if(args[0] is set){
    //try to find player
}elseif(i am a player){
    //use current player as target
}else{
    //invalid syntax
}
```

Commit:
remove useless condition

Commit:
better logic flow for finding target

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,173 @@
+			throw new InvalidCommandSyntaxException();
+		}
+
+		if($target === null){
```
This block could be moved inside the block which calls Server->getPlayer().

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,173 @@
+					$target->getArmorInventory()->clear($index);
+				}
+
+				foreach($target->getInventory()->all($item) as $index => $i){
```
An underflow will occur here if only 1 item wanted to be cleared and it was already removed from the armor inventory.

Commit:
move check to where `Server->getPlayer()` is called

Commit:
fix a possible underflow

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,174 @@
+		$maxCount = -1;
+		if(isset($args[2])){
+			$item->setCount($maxCount = $this->getInteger($sender, $args[2], 0));
+		}
```
You can move this block inside the previous one to fix the PHPStan error.

Ifera replied:
Done

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,174 @@
+					$target->getArmorInventory()->clear($index);
+				}
+
+				foreach($target->getInventory()->all($item) as $index => $i){
```
`$index` already exists outside the loop, so PHPStan doesn't like you overwriting it. You can use a different variable name to get around it.

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,174 @@
+				}
+
+				if($maxCount > 0){
+					foreach($target->getInventory()->all($item) as $index => $i){
```
Same as above.

Commit:
move max count check inside the try...catch... block

Commit:
fix phpstan reported error

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,174 @@
+			}
+
+			if($count > 0){
+				Command::broadcastCommandMessage($sender, new TranslationContainer("%commands.clear.testing", [$target->getName(), $count]));
```
This doesn't need to be a `broadcast` since it doesn't actually do anything, it's only viewing contents.

dktapps commented on src/command/defaults/ClearCommand.php:
```diff
@@ -0,0 +1,174 @@
+
+		//clear everything from the targets inventory
+		if($item === null){
+			$contents = array_merge($target->getInventory()->getContents(), $target->getArmorInventory()->getContents());
```
Count of cursor inventory is missing here.

dktapps replied:
I'm not sure if `clear` is intended to clear the cursor or not.

Ifera replied:
Does the count of cursor inventory matter tho?

dktapps replied:
Yes, I would think so. Either don't clear it, or include it in the total.

Ifera replied:
I opted, not to clear it.

Commit:
no need for broadcast

Commit:
dont clear cursor inv

dktapps committed:
change TODO format

this makes it easier to grep for TODOs.

Commit:
Merge remote-tracking branch 'upstream/master' into master

Commit:
fix typo

dktapps committed:
Enhance type information in PlayerCreationEvent

dktapps committed:
[ci skip] update crash issue template

dktapps committed:
Player: Improved hardcore death handling
it's now possible to cancel the banning by cancelling PlayerKickEvent.
In addition, the disconnect message will now properly explain why the player is banned instead of just saying 'You have been banned',
and also banned-players.txt will show 'Died in hardcore mode' as the reason instead of being completely blank (seriously, we couldn't tell the difference between real bans and hardcore bans????)

dktapps committed:
[ci skip] updated changelog

dktapps committed:
PermissionManager: fixed nonsensical PHPDoc type

dktapps committed:
Permission: remove dead code
Bukkit CCC at its finest... what the fuck this was doing anyway I have no idea
an addChild() method would have made far more sense, especially considering that addParent() doesn't actually mutate the child.

dktapps committed:
DefaultPermissions: remove ridiculous recursion
this line is totally useless since the recursion does exactly the same thing as if it wasn't there.

dktapps committed:
Do not mutate the result of Permission->getChildren()

dktapps committed:
Permission: added removeChild()
this probably isn't useful at all, but it's best to provide alternatives for all the stuff that getChildren() could have been used for ...

dktapps committed:
Permission->getChildren() no longer returns by-ref

dktapps committed:
Removed Server->broadcast()
this method has an explicit dependency on permission subscriptions, which I plan to eradicate.
The method's functionality can be replaced using Server->broadcastMessage() with a custom recipients array.

dktapps committed:
Strip some junk out of IPlayer
these methods have better pathways through Server directly.
Also, setBanned() does not allow customising the reason for banning or the duration, nor does isBanned() account for IP bans because the code is so old ... better to force dependence on a central code path to avoid these inconsistencies.
I want to do the same thing for OP, but that's a separate problem due to its effects on the permission system.

dktapps committed:
PermissionAttachment: better document the apparently nonsensical code responsible for making sure the entire permission system doesn't burn down

dktapps committed:
Merge branch 'stable'

# Conflicts:
#	resources/vanilla
#	src/event/player/PlayerCreationEvent.php

dktapps committed:
Toss ServerOperator; PermissibleBase now tracks its own op status
as well as being simpler, this also allows some nice features, such as granting memory-only op state that goes away when the player quits the server.

dktapps committed:
ZippedResourcePack: improved exception error on invalid manifest contents

dktapps committed:
[ci skip] changelog updates

dktapps committed:
PermissibleBase: fixed root permissible being unaware of changes in some cases
for example, Player->recalculatePermissions() was not called when a player's op status changed, nor when a permission attachment altered permissions (potential secvuln for broadcasting).

dktapps committed:
Rewrite message broadcasting system to not depend on PermissionManager subscriptions
relying on permission subscriptions for this was unreliable (a permissible is not always subscribed to a permission even when it does have it), and also difficult to control (for example there have been various bugs in the past where a Player ended up subscribed to broadcast permissions when it didn't expect to be, thanks to permission recalculation happening too early).
In addition, we might in the future want to have broadcast receivers which are not permissibles (i.e. a more general interface than CommandSender (why does a broadcast receiver need to also be a command sender, anyway?)), which the permission system wouldn't be suitable for.

dktapps committed:
[ci skip] reflect today's changes to Server in changelog

dktapps committed:
PermissionParser: remove unmaintained code

dktapps committed:
OfflinePlayer: remove more useless junk
to obtain an OfflinePlayer instance you had to have a server instance to start with, which means getServer() is pointless, and also that isOp() and setOp() (which can be removed thanks to the fact that OfflinePlayer is not a Permissible) can be done by just asking the server.

dktapps committed:
OfflinePlayer no longer depends on Server
it also doesn't provide stuff like isOnline() (you had to have a Server reference to get an OfflinePlayer anyway, just ask the Server if the player is online ...) and getPlayer().

dktapps committed:
PlayerDataSaveEvent: provide online player if possible, otherwise don't load offline data from disk for no reason (we already have it, that's the whole point of the event ...)

dktapps committed:
Permissions are now always false when permission calculation hasn't been done yet
lack of permission calculation means that child permissions might not have been set correctly, so this might lead to users being able to access things they aren't supposed to.

dktapps committed:
NetworkStackLatencyPacket: added named constructors

dktapps committed:
Sync composer dependencies

dktapps committed:
phpstan 0.12.58

DerDevHD committed:
Fixed crash when executing command /version with multiple authors declared in 'author' (#3940)

closes #3902

dktapps committed:
Removal of permission defaults (in favour of permission cascading) (#3937)

dktapps committed:
[ci skip] update changelog to reflect permission changes

dktapps committed:
PermissionAttachmentInfo::__construct() never throws an exception

dktapps committed:
phpunit 9.4.4

dktapps committed:
Merge branch 'stable'

# Conflicts:
#	composer.lock
#	resources/vanilla
#	src/permission/BanEntry.php
#	src/permission/PermissionAttachmentInfo.php

dktapps committed:
PermissionAttachmentInfo no longer references Permissible
since the only way to obtain PermissionAttachmentInfo via the API is from the Permissible API, it makes zero sense to ask the attachmentinfo for its permissible, since we obviously already had it to be able to get the info in the first place. Therefore, this is just another useless reference stopping Permissibles from being garbage-collected.

dktapps committed:
Timings: remove obsolete timings handler

dktapps committed:
PermissibleBase: more robust cycle destruction

dktapps committed:
Implemented a better method for detecting permission recalculation
this allows anyone to listen to permissions being recalculated, which is useful for stuff like broadcast channel subscriptions.

dktapps committed:
Living: wipe out effect add/remove hooks onDispose()

dktapps committed:
PermissibleBase no longer circularly references its parent permissible
this was used for 2 things:
- to notify parent permissibles on recalculate permissions (not needed anymore thanks to callbacks)
- subscription to chat broadcast permissions (now handled separately, again using callbacks)

therefore, it's no longer needed for PermissibleBase to reference its parent, meaning that we can have one less circular dependency in our lives.

dktapps committed:
Player: added API method getUsedChunks()

dktapps committed:
RakLibServer: do not try to export exceptions across threads (they might contain Threaded objects whose references get invalidated)

dktapps committed:
custom threads: use PTHREADS_INHERIT_NONE by default
this has the most reliable behaviour, because some stuff like respect/validation likes to include files unconditionally which declare functions that are already declared, which breaks due to the lack of function autoloading.

dktapps committed:
Improve painting performance
these really shouldn't hit the CPU at all considering they are simply static objects ...

dktapps committed:
resource packs: added metadata to manifest models

dktapps committed:
Remove useless group permission nodes
the only value these have is for convenience, and a permission manager plugin can easily generate them by parsing the permission IDs.

dktapps committed:
PermissibleBase: compute permission diff, do not fire callbacks when diff is empty
this fixes AvailableCommandsPacket getting sent twice when adding a PermissionAttachment.

dktapps committed:
Permissible->recalculatePermissions() now returns a map of changed permissions with their old values

dktapps committed:
PermissibleBase: fixed express denying of a root permission not working
this only took effect when a permission was applied as a root. When overridden with a permission attachment, the problem did not manifest.

dktapps committed:
PermissibleBase: allow injecting more dynamic base permissions

dktapps committed:
PermissionAttachmentInfo now includes a reference to the PermissionAttachmentInfo that caused its assignment (if any)
this allows plugins to easily identify which permission group caused a permission to be assigned (particularly if multiple groups might cause it).

dktapps committed:
PermissibleBase: simplify diff calculation

dktapps committed:
Introduced pocketmine.group.console
this is only used for /dumpmemory right now, but it's planned to be used for more stuff.

dktapps committed:
Updated DevTools submodule to pmmp/DevTools@7c2d05cafc52328f57b50daab5d58672501baad6

dktapps committed:
PermissibleBase: do not assign any permissions by default
there is no longer such a thing as a permission granted to 'everyone', since we're no longer limited to true/false/op/notop, and regular players are now assigned the pocketmine.group.user permission.
It's possible we might want to add more restricted groups than 'user' in the future, in which case it would be behaviour-BC-breaking to change the default assigned permissions when creating a new PermissibleBase. Therefore, it's better to not assign any permissions at all and let the caller decide.
In addition, this solves the problem of implied permission subscriptions and leak on PermissibleBase construction - if base permissions are provided, it should be expected that the permissible will subscribe to those permissions.

dktapps committed:
Network->registerInterface() now returns whether the registration succeeded or not

dktapps committed:
Added a PacketBroadcaster interface
this gives a bit more control over how packets are broadcasted, which might be useful if the batch format changes (e.g. adding a length prefix) for multi version.
This really ought to be unique to a protocol context instead of a network interface, but for now this is the best we can do.

dktapps committed:
Fixed players losing all permissions when being de-opped after joining as op

dktapps committed:
Network: fixed interfaces registered after raw packet handlers not respecting raw packet handler filters

dktapps committed:
SurvivalBlockBreakHandler: fixed block break effects stopping beyond 4 blocks distance
this was actually intended to be 16 blocks ...

dktapps committed:
Use Snooze to improve AsyncTask collection times
regardless of how long an async task takes to run, it will take a multiple of 50ms to get the result processed. This delay causes issues in some cases for stuff like generation, which causes locking of adjacent chunks, and async packet compression, which experiences elevated latency because of this problem.
This is not an ideal solution for packet compression since it will cause the sleeper handler to get hammered, but since it's already getting hammered by every packet from RakLib, I don't think that's a big problem.

dktapps committed:
Separated TimingsRecord from TimingsHandler (different lifetimes)
this fixes #3906.
TimingsHandler will now be automatically destroyed when there are no TimingsRecord referencing it and its owner has no references left to it.
TimingsRecord will be automatically destroyed at the end of the timings session (if active).

This fixes records from temporary timers being lost after the timings handlers are removed, while also fixing TimingsHandlers piling up regardless of whether timings are running or not.
This opens the doors to have more volatile timers, such as per-player/per-session timers, which are destroyed on player quit, but which won't result in the timings disappearing from the final timings result.

dktapps committed:
World: fire ChunkLoadEvent in setChunk() if the chunk did not previously exist
I'm of two minds whether this should continue to fire onChunkChanged() on chunk loaders or not, since the chunks won't be referenced by any chunk loader anyway by virtue of not existing ...

dktapps committed:
World: do not fire ChunkListener->onChunkChanged() when a new chunk is set
the documentation expressly describes a chunk being _replaced_ by a new chunk. This doesn't fit when the chunk didn't exist to begin with.

dktapps committed:
World->populateChunk() no longer causes ChunkLoadEvent to fire with an empty chunk
instead, it will fire when the chunk comes out of PopulationTask and is set into the world using setChunk().
There is still one place left where auto-creation of empty chunks is used by the core, and that's an issue i'm still deciding how to deal with.

dktapps committed:
Entity: flag for despawn if entering ungenerated terrain
previously this created an empty chunk for the entity to hang around in until the terrain was generated, but that's problematic for several reasons, most importantly the fact that non-generated chunks are not saved.

dktapps committed:
World->registerChunkLoader() no longer causes creation of an empty chunk
this only made sense to work around the server crashing when creating a player in ungenerated terrain, but spawning at y=256 is hardly preferable, and it'll crash since fd99445c5bbd2a5a6fdd0109d8a196ea6160e0a3 anyway ...

dktapps committed:
World: remove create params from loadChunk() and loadChunkAtPosition()

dktapps committed:
generation: drop AsyncWorker thread-local storage in favour of a more specialized storage

dktapps committed:
Do not persist SimpleChunkManager between async tasks
this is just asking for memory leaks.

dktapps committed:
LightPopulationTask: remove useless @var

dktapps committed:
FastChunkSerializer no longer encodes chunk coordinates
in cases like PopulationTask it makes more sense to store the coordinates separately where they can be stored more efficiently (once instead of 9 times)
In addition, PopulationTask shouldn't need to serialize an empty chunk just to copy coordinates.

I've made changes like this in other areas already in preparation for the day when chunks no longer contain their coordinates, so this brings us one step closer to that goal.

dktapps committed:
Chunks no longer contain their own coordinates

dktapps committed:
PopulationTask no longer creates chunks if they don't exist
creating a throwaway empty chunk on the main thread is pointless.

dktapps committed:
Chunks no longer exist in un-generated state
a non-generated chunk is now always represented by NULL. This forces the case of ungenerated chunks to be handled by all code, which is necessary because ungenerated chunks cannot be interacted with or modified in any meaningful way.

dktapps committed:
Cleaned out dead error patterns from phpstan level 8 baseline

dktapps committed:
fix PopulationTask crash

HighTecDev committed:
register HellBiome (#3950)

dktapps committed:
Player: fixed crash when stopping using a chunk before it's finished being generated

dktapps committed:
Make sure generator gets preemptively registered when a worker restart is detected
if a PopulationTask took place after the target worker was garbage collected, the population would fail and the chunks it used would be copied for nothing.
This change marks workers as having unregistered generators when detecting that a worker that previously had a generator registered is restarted.

dktapps committed:
World: disallow block placement and breaking in unloaded, ungenerated and locked chunks

dktapps committed:
ConsumingItemAnimation: fixed itemID translation for 1.16.100

dktapps committed:
Added a DedicatedQueryNetworkInterface to ensure Query functionality when RakLibInterface is disabled (#3942)

dktapps committed:
Removed the cycle between Entity and Chunk
it's now the World's responsibility to manage adding/removing entities from appropriate chunks. Entities no longer know or care that chunks exist.
Entity->checkChunks() remains as-is for backwards compatibility - now it just calls the world to sync its position.

dktapps committed:
[ci skip] more world stuff for changelog

dktapps committed:
BaseInventory: fixed passing NULL slots to InventoryListener->onContentChange()
as per the documentation, InventoryListener->onContentChange() does not expect to receive NULL in the given array.

dktapps committed:
Clean up nonsensical code in NetworkBinaryStream->getSlot()

dktapps committed:
Merge branch 'stable'

# Conflicts:
#	resources/vanilla
#	src/network/mcpe/protocol/serializer/PacketSerializer.php
#	src/world/biome/Biome.php
#	tests/phpstan/configs/l8-baseline.neon

dktapps committed:
clean dead errors out of phpstan l8 baseline

dktapps committed:
World: ignore entities calling onEntityMoved() who aren't members of the world
this can happen if movement or teleportation occurs during the creation of an entity.

dktapps committed:
Finalize Utils class

dktapps committed:
World->getHighestBlockAt() may still return -1 if the queried column of blocks is all air

dktapps committed:
ChunkManager->setChunk() no longer accepts NULL

dktapps committed:
PermissionParser: give permissions default FALSE if they don't specify their own default

dktapps committed:
Updated DevTools submodule to pmmp/DevTools@888d021260efa5456cd7cc0174fb3a2663960811

dktapps committed:
Merge branch 'master' of https://github.com/Ifera/PocketMine-MP into clear

dktapps committed:
fix compatibility with d9de775c27109052b987f2feadf714afaf94241e

dktapps commented:
ugh, github... all these extra commits don't show in local `git log`, it's ok...

dktapps committed:
Fixed player not being able to clear their own inventory of a specific item
we only reject this if the target is specified AND is not the player itself.

dktapps committed:
workaround for commands.generic.permission a la other commands
this should really be fixed by adding a new pocketmine. prefixed translation string, but that's a task for later.

dktapps committed:
Updated resources/locale submodule

dktapps commented:
Merged manually as c4b47c6c067528454b685d68af61c705ddd8ce49. Thank you.
